plugin/pkg/admission/gc/gc_admission.go

```diff
@@ -99,10 +125,93 @@ func isChangingOwnerReference(newObj, oldObj runtime.Object) bool {
 	return false
 }
 
+// translates ref to a DeleteAttribute deleting the object referred by the ref.
+func (a *gcPermissionsEnforcement) toDeleteAttribute(ref metav1.OwnerReference, attributes admission.Attributes) (authorizer.AttributesRecord, error) {
+	groupVersion, err := schema.ParseGroupVersion(ref.APIVersion)
+	if err != nil {
+		return authorizer.AttributesRecord{}, err
+	}
+	mapping, err := a.restMapper.RESTMapping(schema.GroupKind{groupVersion.Group, ref.Kind}, groupVersion.Version)
+	if err != nil {
+		return authorizer.AttributesRecord{}, err
+	}
+	return authorizer.AttributesRecord{
+		User: attributes.GetUserInfo(),
+		Verb: "delete",
+		// ownerReference can only refer to an object in the same namespace, so attributes.GetnNamespace() equals to the owner's namespace
+		Namespace:       attributes.GetNamespace(),
+		APIGroup:        groupVersion.Group,
+		APIVersion:      groupVersion.Version,
+		Resource:        mapping.Resource,
+		Name:            ref.Name,
+		ResourceRequest: true,
+		Path:            "",
+	}, nil
+}
+
+// only keeps the blocking refs
+func blockingOwnerRefs(refs []metav1.OwnerReference) []metav1.OwnerReference {
+	var ret []metav1.OwnerReference
+	for _, ref := range refs {
+		if ref.BlockOwnerDeletion != nil && *ref.BlockOwnerDeletion == true {
+			ret = append(ret, ref)
+		}
+	}
+	return ret
+}
+
+func indexByUID(refs []metav1.OwnerReference) map[types.UID]metav1.OwnerReference {
+	ret := make(map[types.UID]metav1.OwnerReference)
+	for _, ref := range refs {
+		ret[ref.UID] = ref
+	}
+	return ret
+}
+
+// Returns new blocking ownerReferences, and references whose blockOwnerDeletion field is changed.
+// Changes between nil and false are ignored.
+func changingBlockOwnerDeletion(newObj, oldObj runtime.Object) []metav1.OwnerReference {
+	newMeta, err := meta.Accessor(newObj)
+	if err != nil {
+		// if we don't have objectmeta, we don't have the object reference
+		return nil
+	}
+	newRefs := newMeta.GetOwnerReferences()
+
+	if oldObj == nil {
+		return blockingOwnerRefs(newRefs)
+	}
+	oldMeta, err := meta.Accessor(oldObj)
+	if err != nil {
+		// if we don't have objectmeta, treat it as if all the ownerReference are newly created
+		return blockingOwnerRefs(newRefs)
+	}
+
+	var changedRefs []metav1.OwnerReference
+	indexedOldRefs := indexByUID(oldMeta.GetOwnerReferences())
+	for _, newRef := range newRefs {
```

Comment:
if an OwnerRef goes away in `newObj`, this will go unnoticed. Might be intended, but makes this unsymmetric. The func names all sound symmetric (`changingBlockOwnerDeletion`).